content/code-generation-from-the-ast.md

```diff
@@ -0,0 +1,287 @@
++++
+author = ["Ian Eyberg"]
+date = "2015-03-13T12:50:47-07:00"
+linktitle = "Code Generation From the AST"
+title = "Code Generation From the AST"
++++
+
+[Deprehend](https://github.com/deferpanic/deprehend "Deprehend") is a tool
+that auto-generates go src to do things like wrapping goroutines with
+panic handlers and http handlers for metrics tracking.
+
+![](/postimages/code-generation-from-the-ast/deprehend.png)
+
+Over at [DeferPanic](https://deferpanic.com "DeferPanic") we do a lot of
+application monitoring specifically for you, go engineers.
+
+Go as a language is very opinionated and more importantly, go engineers
+are an anti-magic crowd. To this end, we have exposed most of our
+functionality as wrappers around the stdlib. What this means is that you
+have to manually add code to use the client library effectively.
+
+Engineers are a lazy bunch - it's why we are engineers. We like to
+automate as much as possible.
+
+As more and more companies started integrating with us we kept hearing
+more and more of - “You mean I have to wrap every single http handler
+and every single go routine?”
+
+This reminded gophers of their initial reaction to error handling in go
+- "You mean I have to wrap every function with error handling?” :)
+
+For smaller codebases or new projects this isn't a big deal but for
+existing codebases this can be a non-trivial amount of work and what
+happens if you forget something?
+
+We admit, this is not only a complete pain in the ass but it is hard to
+scale as you have no way of knowing new code has been handled correctly
+or not - it kinda defeats the purpose of handling un-expected
+errors/panics.
+
+So we started hacking out Deprehend .
+
+This tool looks at your source code via the AST and auto-generates the
+code for you.
+
+## What exactly is an AST anyways?
+
+AST stands for 'abstract syntax tree'. This is a data structure used
+by the compiler. It performs a number of functions such as looking at
+the correctness of your program and generating the symbol table.
+
+What do we mean by the correctness? Have you ever tried to build your
+program via go build and it fails to compile? Maybe like me you use
+[Fresh](https://github.com/pilu/fresh "Fresh") and you have a OCD of
+hitting save every 2-3 seconds as you are thinking - you'll probably get
+a few failed builds here and there as you save if you have only written
+half an expression. This is the compiler failing to produce a correct
+AST of which it complains and then bails out.
+
+What about the symbol table? The symbol table is another data structure
```

Comment:
That's misconstrued - http://golang.org/pkg/debug/elf/#File.Symbols - they are merely stating that it *can* be parsed without a symbol table - some languages you can't do that in w/out one. There are many other things that have to occur besides parsing.